### ISSUE TITLE
[Feature request] Typeahead tags

### ISSUE THREAD

**Issue by DaAwesomeP** [angular-ui/bootstrap#5316]
How difficult would it be to integrate a multiselect/tag system into Typeahead? This is the feature in a textbox where multiple items are separated into separate boxes. You've probably seen it when you add recipients to an email. There are other tag modules for Angular, but none of them integrate specifically with the Typeahead in ui-bootstrap. I deally they would be represented by an array of object or array of strings, where the object could specify type for color (i.e. danger, warning, etc.), scope attributes for a custom temple, a value to return, and text to display.

I was thinking about doing this myself (unless someone else wants to) and then submitting a pull. Has someone else already done this? Would it be too far from vanilla Bootstrap?

**Comment by icfantv** [comment]
@DaAwesomeP, the angular UI team already has a product for this called ui-select.  You can find it [here](https://github.com/angular-ui/ui-select).

**Comment by DaAwesomeP** [comment]
@icfantv Thanks!

**Comment by shyamal890** [comment]
@DaAwesomeP Hey you should talk to ui-select's owner as he no longer maintains the repository. Check this out https://github.com/angular-ui/ui-select/issues/1398

I wonder why Angular UI guys haven't taken any steps to solve this. May be @icfantv @wesleycho  can share some insights.

**Comment by pkozlowski-opensource** [comment]
> I wonder why Angular UI guys haven't taken any steps to solve this.

@shyamal890 this is open source... If there are no maintainers it probably means that there is no enough interest in a given piece of code (and by interest I don't mean "I just want to use it" but rather "I need it bad enough that I'm willing to spend time contributing / maintaining a given library"

**Comment by shyamal890** [comment]
@pkozlowski-opensource  Hmm, you may be right. Then again, Angular UI team has https://angular-ui.github.io/ mentioning ui-select as a library under their umbrella right?! 

So may be they start thinking on:
Do we really need to have an un-maintained code under our umbrella?

**Comment by icfantv** [comment]
@shyamal890, @ProLoser was looking to offload the project last summer but no one was interested.  He had some ideas about how to restructure the code to make it more dependent on the end user rather than the library.  With out getting into more politics here, suffice it to say that if the project is no longer maintained then it is no longer being maintained and you'll need to find a replacement.  We're in the same boat at work with respect to several other UI libraries built on angular.

@DaAwesomeP, OSS is really _hard_.  It requires a lot of work by many dedicated people who all do this in addition to their full-time day jobs and _in their spare time_.  It requires an unbelievable level of patience in dealing with continually fresh batches of users whose personalities range from being fully supportive to totally demanding and self-righteous.  It takes a really special person to be able to do this.  If one isn't careful about balance, one gets burned out on OSS and then one quits...sometimes for good.

An OSS project usually begins by someone having or seeing a need and then filling it.  You guys need something and while the pain part is taking someone else's code, learning it, and then figuring out how to make it better, is long and time consuming - your best bet here is probably to help take up the reigns.

Your other alternative is to take a maintained jQuery project and then wrap it.  Which we have to do at work for some of the libraries that are built on Angular but no longer maintained.

As far as tagging support goes within the UIBS project - until Bootstrap adds a mechanism for it - I don't really see us supporting it.  The UIBS project has no CSS and we don't want to start maintaining our own.

**Comment by shyamal890** [comment]
@DaAwesomeP This may be an impetus to you: It seems ng-tags-input too is not maintained anymore. Look at this https://github.com/mbenford/ngTagsInput/issues/610

All in all there is a need for tags support with Angular UI Bootstrap. 

@icfantv 
>As far as tagging support goes within the UIBS project - until Bootstrap adds a mechanism for it - I don't really see us supporting it. The UIBS project has no CSS and we don't want to start maintaining our own.

You do support **datepicker** and **time picker**.  :/

**Comment by icfantv** [comment]
@shyamal890 - touché.  :-)

I would want to see a modularized UIBS first before working on something like this given the ramifications.  I need a maintained alternative for my own job but I don't have the work bandwidth to tackle something like this at the moment.  If we could make it work with the Bootstrap `badge` class that might make things interesting (and I freely admit I've wanted to experiment with that).

Chosen is, by far, the best tagging library out there that I've seen (even better than Select2), but it doesn't allow for free-form tagging and that's a deal breaker for a LOT of people.  It has been requested in the past but last I checked the author(s) have decided to not support that feature - although it's been awhile since then.

**Comment by DaAwesomeP** [comment]
First of all, I didn't mean to start any strife here. However, everyone's made some good points. Really, there needs to be some new, well maintained library.

@icfantv I might just make some directives for Chosen. What do you mean by "free-form tagging?" If you mean having an object with lots of data in each tag, then that might be able to be solved with an auxiliary library that handles the data itself and sends a "display array" to Chosen.

Anyways, since I'm going to make something, I'd be happy to work with other people on it. I need something that works with Typeahead and fits the Bootstrap theme conventions (e.g. uses the badge class).

**Comment by ProLoser** [comment]
Before you begin making yet ANOTHER variation of this plugin why not jump over to ui-select and assist with the rewrite. Then make a bootstrap wrapper for it.

**Comment by DaAwesomeP** [comment]
@ProLoser Sure, I'll just fork it and make a pull when I'm done. What library do you prefer that I base it on (Chosen, Select2, etc.)?

**Comment by shyamal890** [comment]
@DaAwesomeP I think you should support https://github.com/mbenford/ngTagsInput .Check it out

**Comment by ProLoser** [comment]
@DaAwesomeP please refer to this discussion before starting any new work: https://github.com/angular-ui/ui-select/issues/1398#issuecomment-174617767
